[FEATURE] Improved constraint handling, pt 1

nyalldawson opened the pull request:
This PR reworks how the existing attribute constraints are handled and adds some new features:
- constraints (including not null) are now attached to a QgsField (previously was in QgsEditFormConfig). I believe this is a better fit, since there's no reason constraint validation & handling must only be done by the attribute form. Semantically I think it's more appropriate too - a field has a constraint.
- a new constraint for unique values has been added. This tests whether the same value exists for any OTHER feature (ie, ignores the values in the current feature being checked).
- constraints can be automatically detected at the provider level. This is implemented for the postgres and spatialite providers, for the not null and unique constraints.
- users CANNOT remove provider side constraints, but they can add (and remove) additional constraints to a field within QGIS
- adds a new QgsVectorLayerUtils class with static helper methods for working with layers, but which don't necessarily belong in the (already large) QgsVectorLayer class itself.

Part 1 only enforces unique constraint checking within the attribute dialog. Part 2 will bring handling of unique value constraints to map tools which create features (eg split).

nyalldawson committed:
Add method to fetch constraints from a vector data provider

Implemented for unique and not null constraints for postgres
provider

nyalldawson committed:
Add gui for setting/exposing constraints in field properties

Provider set constraints cannot be changed by users, ie
if a not null constraint is set on the field at the database
then QGIS users cannot clear this constraint.

nyalldawson committed:
Move constraints to QgsField

nyalldawson committed:
Allow constraints to be added at a QgsVectorLayer level

nyalldawson committed:
Move handling of not null constraint from edit form to layer

nyalldawson committed:
Select all text after reset line edit to null

nyalldawson committed:
Implement constraint detection for spatialite provider

nyalldawson committed:
New class QgsVectorLayerUtils

Contains static helper methods for working with vector layers.
Initially only contains a method to test whether a value
exists within a layer's attributes.

nyalldawson committed:
Enforce unique constraints in attribute form

nyalldawson committed:
Refactor constraint handling

- store constraint origin in QgsField
- move handling of constraint expressions to QgsField

nyalldawson committed:
Move responsibility for storage of constraint expressions to QgsVectorLayer

nyalldawson committed:
Move responsibility for testing for attribute against constraints
to QgsVectorLayerUtils::validateAttribute()

Also clean up some strings shown to the user when a constraint
check fails

nyalldawson committed:
If the layer is NOT being edited then only check layer based constraints
and not any constraints enforced by the provider

Because:

1. we want to keep browsing features nice and responsive. It's nice to give
 feedback as to whether the value checks out, but not if it's too slow to
 do so. Some constraints (eg unique) can be expensive to test. A user can
 freely remove a layer-based constraint if it proves to be too slow to
 test, but they are unlikely to have any control over provider-side
 constraints

2. the provider has already accepted the value, so presumably it doesn't
 violate the constraint and there's no point rechecking!

nyalldawson committed:
Update API break docs

nyalldawson commented:
@m-kuhn this differs a little from the original proposal - I got a bit carried away with refactoring stuff and cleaning up the existing constraint handling!

nyalldawson committed:
Fix failing unit test

m-kuhn commented:
Thanks Nyall. In general the proposal sounds good, I'll have a look at the code as well.

Something that has recently been identified as requirement are "optional" constraints (silly name?).
I.e. recommend and print warnings but do not enforce setting it inside the form (digitizing features in real life unfortunately sometimes involves incomplete information). Do you think such a flag could be saved into QgsField as well or should that be kept on form level since it's about gui hints only?

nyalldawson commented:
What types of optional constraints are these? Unique/not null or expressions? (or all?)

m-kuhn commented:
In most use-cases probably not null (but it might happen to others as well). Why do you mean?

m-kuhn commented on src/core/qgsfield.h:
```diff
@@ -208,6 +233,65 @@ class CORE_EXPORT QgsField
+     * @see constraints()
+     * @see removeConstraint()
+     */
+    void setConstraint( Constraint constraint, ConstraintOrigin origin = ConstraintOriginLayer );
```
I wonder if that should be public API like that or of we want to make that private and have a public

    void setConstraint( Constraint constraint );

instead.
Then make QgsVectorDataProvider a friend class so it can still set the private constraints.

m-kuhn commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -1449,6 +1449,44 @@ bool QgsVectorLayer::readXml( const QDomNode& layer_node )
     }
   }
+
+  // constraints
```
I didn't check... Is this backwards compatible?

m-kuhn commented:
Good work - as usual. Only minor questions inside.

m-kuhn commented:
@pblottiere any feedback from your side?

palmerj commented:
It would be good to implement provider side null/not null constraints for ogr layers as well. GDAL/OGR has supported this since 2.0

m-kuhn commented on src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp:
```diff
@@ -113,61 +113,25 @@ void QgsEditorWidgetWrapper::updateConstraint( const QgsFeature &ft )
   if ( field.constraints() & QgsField::ConstraintNotNull )
   {
-    descriptions << QStringLiteral( "NotNull" );
+    descriptions << QStringLiteral( "Not NULL" );
```
Should that be `tr()`anslated?

m-kuhn commented on tests/src/gui/testqgsattributeform.cpp:
```diff
@@ -78,8 +78,8 @@ void TestQgsAttributeForm::testFieldConstraint()
-  QString validLabel = QStringLiteral( "col0<font color=\"green\">*</font>" );
-  QString invalidLabel = QStringLiteral( "col0<font color=\"red\">*</font>" );
+  QString validLabel = QStringLiteral( "col0<font color=\"green\">✔</font>" );
+  QString invalidLabel = QStringLiteral( "col0<font color=\"red\">❌</font>" );
```
I'd prefer ✘ ;)

nyalldawson replied:
done!

nyalldawson commented:
@m-kuhn the other approach I was thinking of taking (but was concerned it was overkill) would be to have a class QgsConstraint, and make QgsField take/return lists of QgsConstraints. Maybe this would work better for "soft" (non enforced) constraints.

What I *don't* want to do is make QgsConstraint an abstract class... I'm pretty sure that's overengineering ;)

What do you think?

m-kuhn commented:
Ah, I see now why you were asking for the type of constraint.

Another option would be a field-level flag for "enforce constraint", that wouldn't require the constraints class and be an easy win.

For more fine-grained control we could also add a feature-level list of constraints in the future.

nyalldawson commented:
I'm hesitant to add another constraint flag into QgsField as I'm concerned it'll clutter the API too much. Options I see are
- move all the constraints stuff to a seperate QgsFieldConstaints class and have a getter/setter for a single QgsFieldConstraints in QgsField
- or the QgsConstraint approach mentioned earlier.

I'm ok with either approach. I suspect we'll have a requirement for foreign key constraints at some stage, so I want to leave room for this. Which approach do you prefer?

m-kuhn commented:
API-wise, a QgsFieldConstraints class sounds most clean to me. But no strong opinion.

nyalldawson committed:
[FEATURE] Not null constraint detection for ogr provider

Implements not null constraint detection for the OGR layers, where
supported for the data format by OGR.

(only available for GDAL >= 2.0)

nyalldawson committed:
Color a few pixels slightly differently

nyalldawson committed:
Move constraint handling to QgsFieldConstraints

Avoids cluttering QgsField API

nyalldawson commented:
@palmerj - done, thanks for the heads up!

nyalldawson commented:
@m-kuhn 
> API-wise, a QgsFieldConstraints class sounds most clean to me. But no strong opinion.

ok, done!

nyalldawson committed:
Fix some untranslatable strings

nyalldawson committed:
fix test

nyalldawson committed:
[FEATURE] Field constraints can be enforced or not

Non-enforced constraints just show a warning to the user, but
do not prevent committing the feature. Enforced constraints
block users from comitting non compliant features.

Any constraints detected by the provider are always
enforced.

nyalldawson committed:
Respect non-enforced constraints when editing/adding features

Warnings are shown, but features can be committed. Fields which
fail an unenforced constraint are now shaded in yellow to differentiate
from the red failure for enforced constraints.

nyalldawson commented:
@m-kuhn phew... should be good to go now. Now handles enforced/non-enforced constraints.

nyalldawson committed:
Update api break docs

nyalldawson commented:
Thanks for the reviews @m-kuhn !
